nav2_costmap_2d/plugins/costmap_filters/costmap_filter.cpp

```diff
@@ -0,0 +1,153 @@
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the <ORGANIZATION> nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+#include "nav2_costmap_2d/costmap_filters/costmap_filter.hpp"
+
+#include <exception>
+
+namespace nav2_costmap_2d
+{
+
+CostmapFilter::CostmapFilter()
+: semantic_info_topic_("/semantic_info")
+{
+}
+
+CostmapFilter::~CostmapFilter()
+{
+}
+
+void CostmapFilter::onInitialize()
+{
+  costmap_ = nullptr;
+  costmap_size_x = 0;
+  costmap_size_y = 0;
+
+  // Get parameters
+  declareParameter("enabled", rclcpp::ParameterValue(true));
+  declareParameter("semantic_info_topic", rclcpp::ParameterValue("/semantic_info"));
+
+  node_->get_parameter(name_ + "." + "enabled", enabled_);
+  node_->get_parameter(name_ + "." + "semantic_info_topic", semantic_info_topic_);
+}
+
+void CostmapFilter::activate()
+{
+  loadFilter(semantic_info_topic_);
+}
+
+void CostmapFilter::deactivate()
+{
+  unloadFilter();
+}
+
+void CostmapFilter::reset()
+{
+  if (costmap_) {
+    delete[] costmap_;
+  }
+  costmap_ = nullptr;
+  costmap_size_x = 0;
+  costmap_size_y = 0;
+
+  unloadFilter();
+  loadFilter(semantic_info_topic_);
+}
+
+void CostmapFilter::updateBounds(
+  double robot_x, double robot_y, double robot_yaw,
+  double * /*min_x*/, double * /*min_y*/, double * /*max_x*/, double * /*max_y*/)
+{
+  latest_robot_x = robot_x;
+  latest_robot_y = robot_y;
+  latest_robot_yaw = robot_yaw;
+}
+
+void CostmapFilter::updateCosts(
+  nav2_costmap_2d::Costmap2D & master_grid,
+  int min_i, int min_j, int max_i, int max_j)
+{
+  // (Re)Allocate costmap_ for using in costmap filters
+  reAllocateCostmap(master_grid);
+
+  // Clear costmap_ for process()
+  clearCostmap(min_i, min_j, max_i, max_j);
+
+  process(
+    master_grid, min_i, min_j, max_i, max_j,
+    latest_robot_x, latest_robot_y, latest_robot_yaw);
+}
+
+void CostmapFilter::reAllocateCostmap(
```

Comment:
I don't understand why this is needed or why its being implemented here when there are other costmap_2d / layer implementations of similar things like resetMaps(). but even then you're deleting the maps every iterations which is massively costly. Why would any of this be necessary?

Reply:
Manual work with `costmap_` was fixed to using standard Costmap2D API: `matchSize()`, `resetMaps()` and `updateOrigin()`. These methods are enough to make all I've initially made manually.